Move numReadThreads to be a class member so that it can be set extern…

rolandhill opened the pull request:
Make numReadThreads a member of DatabasePager so that it can be set externally if desired. It used to be hardwired in the start() method. The default value is kept the same so there should be no effect on existing code.

robertosfield commented:
The problem with setting the value this way is once the database pager has started the number of read threads will be fixed so subsequent changes to the numReadThreads won't have an affect.

In you app what is your usage pattern?  Perhaps we can come up with a way that allows you the controlability you need without having a number member that only has an effect in one particular usage patter.

rolandhill commented:
At the moment I only intend to change the number of threads once as the
application starts. I hadn't really thought about adjusting it on-the-fly.
I do a reasonable amount of processing when a database tile/octree cell is
requested and the queue can get up over 900 requests pending, which takes
time to clear. Experimenting to use 8 read threads looks to help, at least
on my machine. Ideally I could set it according to the user's CPU resources.

More generally, my users need to be able to change their 'project' at any
time so I've derived my own DBPager from your DatabasePager and included my
own functions for start(), stop(), flush(), updateSceneGraph(). When they
change project I stop the DatabasePager, flush it and then restart. This
also means that I can create my DBPager manually (without it starting), set
the number of threads, then start it. There is nothing in my scene graph
when I call the first compile(), so it doesn't get set up automatically.
All content is added dynamically after the frame loop is running.

If you can think of a better way then I'm happy to go with it.


On Wed, 21 Aug 2024 at 16:56, Robert Osfield ***@***.***>
wrote:

> The problem with setting the value this way is once the database pager has
> started the number of read threads will be fixed so subsequent changes to
> the numReadThreads won't have an affect.
>
> In you app what is your usage pattern? Perhaps we can come up with a way
> that allows you the controllability you need without having a number member
> that only has an effect in one particular usage patter.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/1264#issuecomment-2301275907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPOEQ7NGQOD45AXWQ26DHDZSQ22HAVCNFSM6AAAAABM3C6CT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBRGI3TKOJQG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

rolandhill commented:
It's just after - see below. I can set the number of read threads in the
DbPager constructor with the changes in this PR. Nodes attached at the time
compile() is called are minimal.

_viewer->compile();

// Install the custom database pager
_dbPager = DbPager::create();
_dbPager->compileManager = vsg::CompileManager::create(*_viewer,
resourceHints);
for(auto& task : _viewer->recordAndSubmitTasks)
{
task->databasePager = _dbPager;
}
_dbPager->start();

On Wed, 21 Aug 2024 at 18:39, Robert Osfield ***@***.***>
wrote:

> How are you going about setting up the DatabasePager? Before or after the
> initial Viewer::compile(..) call?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/1264#issuecomment-2301487845>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPOEQ6BJA6AWFIEQKTOZILZSRG4BAVCNFSM6AAAAABM3C6CT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBRGQ4DOOBUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robertosfield commented:
Perhaps we could pass in the settings for the DatabaserPager via the viewer::compile(resourceHints);  So add the number of threads to use into the resourceHints.  This way you wouldn't need any database pager setup code or need to create a new CompileManager or DatabasePager.

rolandhill commented:
I'm happy to try that. As long as the number of threads isn't hard coded as
a local variable in the start() method, then it would be fully flexible for
anyone who wants to customise it.

On Thu, 22 Aug 2024 at 22:04, Robert Osfield ***@***.***>
wrote:

> Perhaps we could pass in the settings for the DatabaserPager via the
> viewer::compile(resourceHints); So add the number of threads to use into
> the resourceHints. This way you wouldn't need any database pager setup code
> or need to create a new CompileManager or DatabasePager.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/1264#issuecomment-2304500645>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPOEQ6K25U5ZBZXJWRNML3ZSXHU3AVCNFSM6AAAAABM3C6CT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBUGUYDANRUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robertosfield commented:
My work on utilizing TransferTask within the CompileTraversal I've ended working on DatabasrePager to fix the occasional high DatabaserPager::updateSceneGraph() times that was causing big frame drops, this was resolved by introducing a dedicated DeleteQueue that now has a delete thread in DatabasePager.  This work's now done, and as I was already working on DatabasrPager I've implemented support for specifying the number of read threads to create with commit: c2a2559f94e3a21ec3f7d7c7e197c98cd9cd7efc

This can be found in the Compile_Using_TransferData branch:  https://github.com/vsg-dev/VulkanSceneGraph/tree/Compile_Using_TransferData

I will do some more testing, if that all goes well I'll merge with VSG master tomorrow.

With these changes you can now do databasePager->start(numReadThreads)l, and as the vsg::Viewer::compile() calls databasePager->start(..) for users I've added a ResourceHints::numDatabasePagerReadThreads member that defaults to 4, you can pass this in the form viewer->compile(hints);

I have checked in some changes to vsgviewer in vsgExampes to enable users to read and output a ResourceHints file using --rh resource.vsgt or --orh resource.vsgt command line options: 

https://github.com/vsg-dev/vsgExamples/commit/b9f3c903d2cfdcc31a9275f53ab3529663f6a5c0

rolandhill commented:
Hi Robert,

I tried the Compile_Using_TransferData branch on Ubuntu 24.04 and setting
the threads using start(8). No problems - all went as expected.

Cheers,
Roland


On Wed, 28 Aug 2024 at 03:03, Robert Osfield ***@***.***>
wrote:

> My work on utilizing TransferTask within the CompileTraversal I've ended
> working on DatabasrePager to fix the occasional high
> DatabaserPager::updateSceneGraph() times that was causing big frame drops,
> this was resolved by introducing a dedicated DeleteQueue that now has a
> delete thread in DatabasePager. This work's now done, and as I was already
> working on DatabasrPager I've implemented support for specifying the number
> of read threads to create with commit: c2a2559
> <https://github.com/vsg-dev/VulkanSceneGraph/commit/c2a2559f94e3a21ec3f7d7c7e197c98cd9cd7efc>
>
> This can be found in the Compile_Using_TransferData branch:
> https://github.com/vsg-dev/VulkanSceneGraph/tree/Compile_Using_TransferData
>
> I will do some more testing, if that all goes well I'll merge with VSG
> master tomorrow.
>
> With these changes you can now do databasePager->start(numReadThreads)l,
> and as the vsg::Viewer::compile() calls databasePager->start(..) for users
> I've added a ResourceHints::numDatabasePagerReadThreads member that
> defaults to 4, you can pass this in the form viewer->compile(hints);
>
> I have checked in some changes to vsgviewer in vsgExampes to enable users
> to read and output a ResourceHints file using --rh resource.vsgt or --orh
> resource.vsgt command line options:
>
> ***@***.***
> <https://github.com/vsg-dev/vsgExamples/commit/b9f3c903d2cfdcc31a9275f53ab3529663f6a5c0>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/VulkanSceneGraph/pull/1264#issuecomment-2313089772>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPOEQZYMJGUH4NNU6Z3GMLZTSWNFAVCNFSM6AAAAABM3C6CT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJTGA4DSNZXGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robertosfield commented:
Thanks for the testing.  I'll close this PR as it looks like the feature required is working fine.
